src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala

```diff
@@ -168,46 +168,64 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
     }
   }
 
-  /**
-   * Read until the given data are matched with `mat`.
-   * When withinBlock is true, it saves the data came in.
-   *
-   * @param mat bytes to match
-   * @param withinBlock start offset
-   * @return whether it finds the match successfully
-   */
-  private def readUntilMatch(mat: Array[Byte], withinBlock: Boolean): Boolean = {
-    var i: Int = 0
-    while(true) {
-      val b: Int = in.read
-      if (b == -1) {
-        currentStartTag = startTag
-        return false
-      }
-      if (withinBlock) {
-        buffer.write(b)
-      }
-      if (b == mat(i)) {
-        i += 1
-        if (i >= mat.length) {
-          currentStartTag = startTag
-          return true
+  private def readUntilStartElement(): Boolean = {
+    @annotation.tailrec
+    def loop(i: Int): Boolean = {
```

Comment:
I could avoid tail recursion and use a while loop instead, but missing cases could only be caught at runtime (with tests) as opposed to compile time (with the compiler). I would also need to use mutable variables and return statements both of which break referential transparency making the code harder to reason about and refactor. In short, I would recommend against it. That said, I appreciate this is a DataBricks repository and their style guide explicitly recommends against tail recursion:

> Avoid using recursion, unless the problem can be naturally framed recursively (e.g. graph traversal, tree traversal).

With the above in mind, would you still like me to rewrite using a while loop?

Regarding comments, I am happy to put those in whether using tail recursion or a while loop. I know some people don't like them, but the code is quite low level, so I agree they will help. Do you prefer above or to the right (subject to line length)?